File: tests/node-bindgen/ts/json.ts
import { Group } from "./common";
import { StructCustomData, EnumA } from "binding";

const tests = new Group("Custom Data Tests");
const struct = new StructCustomData();

{
    const test = tests.test("getData");
    const result = struct.getData({ a: 1, b: 2, c: "test" });
    if (result instanceof Error) {
        test.fail(
            `Function getData() of StructCustomData returns error: ${result.message}`
        );
    } else {
        test.assert(result.a).msg("Value of result.a invalid").equal(2);
        test.assert(result.b).msg("Value of result.b invalid").equal(3);
        test.assert(result.c)
            .msg("Value of result.s invalid")
            .equal("testtest");
        test.success();
    }
}

{
    const test = tests.test("getDataA");
    const result = struct.getDataA({ a: 1, b: 2, c: "test" });
    if (result instanceof Error) {
        test.fail(
            `Function getDataA() of StructCustomData returns error: ${result.message}`
        );
    } else {
        test.assert(result.a.a).msg("Value of result.a.a invalid").equal(2);
        test.assert(result.a.b).msg("Value of result.a.b invalid").equal(3);
        test.assert(result.a.c)
            .msg("Value of result.a.c invalid")
            .equal("testtest");
        test.assert(result.b instanceof Array)
            .msg("Value of result.b invalid")
            .beTrue();
        test.assert(result.b.length).msg("Value of result.b invalid").equal(2);
        test.assert(result.b[0].a)
            .msg("Value of result.b[0].a invalid")
            .equal(2);
        test.assert(result.b[0].b)
            .msg("Value of result.b[0].b invalid")
            .equal(3);
        test.assert(result.b[0].c)
            .msg("Value of result.b[0].c invalid")
            .equal("testtest");
        test.assert(result.b[1].a)
            .msg("Value of result.b[1].a invalid")
            .equal(2);
        test.assert(result.b[1].b)
            .msg("Value of result.b[1].b invalid")
            .equal(3);
        test.assert(result.b[1].c)
            .msg("Value of result.b[1].c invalid")
            .equal("testtest");
        test.assert(result.c).msg("Value of result.c invalid").type("object");
        test.assert(result.c["first"])
            .msg("Value of result.c invalid")
            .typeNot("undefined");
        test.assert(result.c["first"].a)
            .msg("Value of result.b[1].a invalid")
            .equal(2);
        test.assert(result.c["first"].b)
            .msg("Value of result.b[1].b invalid")
            .equal(3);
        test.assert(result.c["first"].c)
            .msg("Value of result.b[1].c invalid")
            .equal("testtest");
        test.assert(result.c["second"])
            .msg("Value of result.c invalid")
            .typeNot("undefined");
        test.assert(result.c["second"].a)
            .msg("Value of result.b[1].a invalid")
            .equal(2);
        test.assert(result.c["second"].b)
            .msg("Value of result.b[1].b invalid")
            .equal(3);
        test.assert(result.c["second"].c)
            .msg("Value of result.b[1].c invalid")
            .equal("testtest");
        test.assert(result.d).msg("Value of result.d invalid").equal("test");
        test.success();
    }
}

{
    const test = tests.test("getMultipleData");
    const result = struct.getMultipleData(
        { a: 1, b: 2, c: "test" },
        { a: 2, b: 3 }
    );
    if (result instanceof Error) {
        test.fail(
            `Function getMultipleData() of StructCustomData returns error: ${result.message}`
        );
    } else {
        test.assert(result[0]).msg("Value of result.a invalid").equal(3);
        test.assert(result[1]).msg("Value of result.b invalid").equal(5);
        test.success();
    }
}

{
    const test = tests.test("getEnumA");
    console.log(EnumA.One);
    const one = struct.getEnumA(EnumA.One);
    const two = struct.getEnumA(EnumA.Two);
    const three = struct.getEnumA(EnumA.Three);
    if (
        one instanceof Error ||
        two instanceof Error ||
        three instanceof Error
    ) {
        test.fail(`Function getEnumA() of StructCustomData returns error`);
    } else {
        test.assert(one).msg("Value of one invalid").equal(EnumA.One);
        test.assert(two).msg("Value of two invalid").equal(EnumA.Two);
        test.assert(three).msg("Value of three invalid").equal(EnumA.Three);
        test.success();
    }
}

{
    const test = tests.test("getEnumB");
    const one = struct.getEnumB({
        One: "test",
    });
    const two = struct.getEnumB({
        Two: [1, 2],
    });
    const three = struct.getEnumB({
        Three: 1,
    });
    const fourSome = struct.getEnumB({
        Four: 1,
    });
    const fourNone = struct.getEnumB({
        Four: null,
    });
    if (
        one instanceof Error ||
        two instanceof Error ||
        three instanceof Error ||
        fourSome instanceof Error ||
        fourNone instanceof Error
    ) {
        test.fail(`Function getEnumB() of StructCustomData returns error`);
    } else {
        test.assert(one.One).msg("Value of one invalid").equal("test");
        test.assert(Object.keys(one).length)
            .msg("Value of one invalid")
            .equal(1);
        test.assert((two.Two as any)[0])
            .msg("Value of two invalid")
            .equal(1);
        test.assert((two.Two as any)[1])
            .msg("Value of two invalid")
            .equal(2);
        test.assert(Object.keys(two).length)
            .msg("Value of one invalid")
            .equal(1);
        test.assert(three.Three).msg("Value of three invalid").equal(1);
        test.assert(Object.keys(three).length)
            .msg("Value of three invalid")
            .equal(1);
        test.assert(fourSome.Four).msg("Value of fourSome invalid").equal(1);
        test.assert(Object.keys(fourSome).length)
            .msg("Value of fourSome invalid")
            .equal(1);
        test.assert(fourNone.Four).msg("Value of fourNone invalid").equal(null);
        test.assert(Object.keys(fourNone).length)
            .msg("Value of fourNone invalid")
            .equal(1);
        test.success();
    }
}

{
    const test = tests.test("testOfErrorSupportOk");
    const result = struct.testOfErrorSupportOk();
    test.assert(result).msg("Value of result invalid").type("number");
    test.assert(result).msg("Value of result invalid").equal(666);
    test.success();
}

{
    const test = tests.test("testOfErrorSupportErr");
    const result = struct.testOfErrorSupportErr();
    if (result instanceof Error) {
        test.assert(result.err)
            .msg("Value of result.err invalid")
            .type("object");
        test.assert(result.err?.code)
            .msg("Value of result.err?.code invalid")
            .equal(666);
        test.assert(result.err?.msg)
            .msg("Value of result.err?.msg invalid")
            .equal("test");
        test.assert(result.err?.err.code)
            .msg("Value of result.err?.err.code invalid")
            .equal(666);
        test.assert(result.err?.err.msg)
            .msg("Value of result.err?.err.code invalid")
            .equal("Error");
    } else {
        test.fail("Value of result invalid");
    }
    test.success();
}

File: tests/node-bindgen/rs/json.rs
use node_bindgen::derive::node_bindgen;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tslink::tslink;

#[tslink]
#[derive(Serialize, Deserialize)]
struct ErrorA {
    msg: String,
    code: usize,
}

impl From<serde_json::Error> for ErrorA {
    fn from(value: serde_json::Error) -> Self {
        ErrorA {
            msg: value.to_string(),
            code: 1,
        }
    }
}

#[tslink]
#[derive(Serialize, Deserialize)]
struct ErrorB {
    msg: String,
    code: usize,
    err: ErrorA,
}

impl From<serde_json::Error> for ErrorB {
    fn from(value: serde_json::Error) -> Self {
        ErrorB {
            msg: value.to_string(),
            code: 1,
            err: ErrorA {
                msg: "test".to_string(),
                code: 1,
            },
        }
    }
}

#[tslink]
#[derive(Serialize, Deserialize)]
struct DataA {
    pub a: i32,
    pub b: i32,
    pub c: String,
}

#[tslink]
#[derive(Serialize, Deserialize)]
struct DataB {
    pub a: DataA,
    pub b: Vec<DataA>,
    pub c: HashMap<String, DataA>,
    pub d: Option<String>,
}

#[tslink]
#[derive(Serialize, Deserialize)]
struct DataC {
    pub a: u32,
    pub b: u32,
}

#[tslink]
#[derive(Serialize, Deserialize)]
enum EnumA {
    One,
    Two,
    Three,
}

#[tslink]
#[derive(Serialize, Deserialize)]
enum EnumB {
    One(String),
    Two(i32, i32),
    Three(i32),
    Four(Option<u8>),
}

struct StructCustomData {}

#[tslink(class)]
#[node_bindgen]
impl StructCustomData {
    #[tslink(constructor)]
    #[node_bindgen(constructor)]
    pub fn new() -> Self {
        Self {}
    }

    #[tslink(
        data = "DataA",
        result = "json",
        error = "json",
        snake_case_naming,
        exception_suppression
    )]
    #[node_bindgen]
    fn get_data(&self, data: String) -> Result<DataA, ErrorA> {
        Ok(DataA {
            a: data.a + 1,
            b: data.b + 1,
            c: format!("{}{}", data.c, data.c),
        })
    }

    #[tslink(
        data = "DataA",
        result = "json",
        error = "json",
        snake_case_naming,
        exception_suppression
    )]
    #[node_bindgen]
    fn get_data_a(&self, data: String) -> Result<DataB, ErrorA> {
        let mut c: HashMap<String, DataA> = HashMap::new();
        c.insert(
            "first".to_string(),
            DataA {
                a: data.a + 1,
                b: data.b + 1,
                c: format!("{}{}", data.c, data.c),
            },
        );
        c.insert(
            "second".to_string(),
            DataA {
                a: data.a + 1,
                b: data.b + 1,
                c: format!("{}{}", data.c, data.c),
            },
        );
        Ok(DataB {
            a: DataA {
                a: data.a + 1,
                b: data.b + 1,
                c: format!("{}{}", data.c, data.c),
            },
            b: vec![
                DataA {
                    a: data.a + 1,
                    b: data.b + 1,
                    c: format!("{}{}", data.c, data.c),
                },
                DataA {
                    a: data.a + 1,
                    b: data.b + 1,
                    c: format!("{}{}", data.c, data.c),
                },
            ],
            c,
            d: Some("test".to_string()),
        })
    }

    #[tslink(
        data_a = "DataA",
        data_c = "DataC",
        error = "json",
        snake_case_naming,
        exception_suppression
    )]
    #[node_bindgen]
    fn get_multiple_data(&self, data_a: String, data_c: String) -> Result<(i32, i32), ErrorA> {
        Ok((data_a.a + data_c.a as i32, data_a.b + data_c.b as i32))
    }

    #[tslink(
        enum_a = "EnumA",
        result = "json",
        error = "json",
        snake_case_naming,
        exception_suppression
    )]
    #[node_bindgen]
    fn get_enum_a(&self, enum_a: String) -> Result<EnumA, ErrorA> {
        Ok(enum_a)
    }

    #[tslink(
        enum_b = "EnumB",
        result = "json",
        error = "json",
        snake_case_naming,
        exception_suppression
    )]
    #[node_bindgen]
    fn get_enum_b(&self, enum_b: String) -> Result<EnumB, ErrorA> {
        Ok(enum_b)
    }

    #[tslink(error = "json", snake_case_naming, exception_suppression)]
    #[node_bindgen]
    fn test_of_error_support_ok(&self) -> Result<i32, ErrorB> {
        Ok(666)
    }

    #[tslink(error = "json", snake_case_naming, exception_suppression)]
    #[node_bindgen]
    fn test_of_error_support_err(&self) -> Result<i32, ErrorB> {
        Err(ErrorB {
            msg: "test".to_string(),
            code: 666,
            err: ErrorA {
                msg: "Error".to_string(),
                code: 666,
            },
        })
    }
}
